pkg/project/auth/watch.go

```diff
@@ -0,0 +1,217 @@
+	// the default `emit` call.  That's why cacheError is a buffered channel.
+	outgoing chan watch.Event
+	// userStop lets a user stop his watch.
+	userStop chan struct{}
+
+	// stopped prevents closing closed channels
+	stopped bool
+	// stopLock keeps parallel stops from doing crazy things
+	stopLock sync.Mutex
+
+	// Injectable for testing. Send the event down the outgoing channel.
+	emit func(watch.Event)
+
+	projectCache *projectcache.ProjectCache
+	authCache    WatchableCache
+
+	knownProjects sets.String
+}
+
+var (
+	watchChannelHWM etcd.HighWaterMark
+)
+
+func NewUserProjectWatcher(username string, groups []string, projectCache *projectcache.ProjectCache, authCache WatchableCache) *userProjectWatcher {
+	userInfo := &user.DefaultInfo{Name: username, Groups: groups}
+	namespaces, _ := authCache.List(userInfo)
+	knownProjects := sets.String{}
+	for _, namespace := range namespaces.Items {
+		knownProjects.Insert(namespace.Name)
+	}
+
+	w := &userProjectWatcher{
+		username: username,
+		groups:   groups,
+
+		cacheIncoming: make(chan watch.Event, 1000),
+		cacheError:    make(chan error, 1),
+		outgoing:      make(chan watch.Event),
+		userStop:      make(chan struct{}),
+
+		projectCache:  projectCache,
+		authCache:     authCache,
+		knownProjects: knownProjects,
+	}
+	w.emit = func(e watch.Event) {
+		w.outgoing <- e
+	}
+	return w
+}
+
+func (w *userProjectWatcher) GroupMembershipChanged(namespaceName string, latestUsers, lastestGroups, removedUsers, removedGroups, addedUsers, addedGroups sets.String) {
+	removed := removedUsers.Has(w.username) || removedGroups.HasAny(w.groups...)
+	added := addedUsers.Has(w.username) || addedGroups.HasAny(w.groups...)
+	hasAccess := latestUsers.Has(w.username) || lastestGroups.HasAny(w.groups...)
+
+	switch {
+	case removed:
+		if !w.knownProjects.Has(namespaceName) {
+			return
+		}
+		w.knownProjects.Delete(namespaceName)
+
+		select {
+		case w.cacheIncoming <- watch.Event{
+			Type:   watch.Deleted,
+			Object: convertNamespace(&kapi.Namespace{ObjectMeta: kapi.ObjectMeta{Name: namespaceName}}),
+		}:
+		default:
+			// remove the watcher so that we wont' be notified again and block
+			w.authCache.RemoveWatcher(w)
+			w.cacheError <- errors.New("delete notification timeout")
+		}
+
+	case added || hasAccess:
+		namespace, err := w.projectCache.GetNamespace(namespaceName)
+		if err != nil {
+			utilruntime.HandleError(err)
+			return
+		}
+		event := watch.Event{
+			Type:   watch.Added,
+			Object: convertNamespace(namespace),
+		}
+
+		// if we already have this in our list, then we're getting notified because the object changed
+		if w.knownProjects.Has(namespaceName) {
+			event.Type = watch.Modified
+		}
+		w.knownProjects.Insert(namespace.Name)
+
+		select {
+		case w.cacheIncoming <- event:
+		default:
+			// remove the watcher so that we won't be notified again and block
+			w.authCache.RemoveWatcher(w)
```

Comment:
> You already defer removing in Watch

`Watch` is not called synchronously.  This remove needs to happen before the next call to `GroupMembershipChanged` or you'll try to write to channel that doesn't have any space and block all callers.  The alternative is to track whether you think you've written an error.  Because locks are not reaquireable, that impl may be more clear.

Also, the go concurrency libraries suck.